tools/c7n_mailer/c7n_mailer/splunk_delivery.py

```diff
@@ -0,0 +1,279 @@
+import json
+from time import sleep
+from six.moves.urllib.parse import urlparse
+from random import uniform
+import requests
+from jsonpointer import resolve_pointer, JsonPointerException
+from jsonpatch import JsonPatch
+from copy import deepcopy
+from .utils import get_aws_username_from_event
+
+
+class SplunkHecDelivery(object):
+    """
+    Delivery class to send c7n message from SQS to Splunk HTTP Event Collector
+    """
+
+    def __init__(self, config, session, logger):
+        """
+        Initialize SplunkHecDelivery HEC sender.
+
+        :param config: sqs_splunk_lambda configuration
+        :type config: dict
+        :param session: boto3 AWS Session
+        :param logger: Logger object to write to
+        """
+        self.config = config
+        self.logger = logger
+        self.session = session
+
+    def get_splunk_payloads(self, msg, msg_timestamp):
+        """
+        Given a raw c7n message dict, generate a list of payloads to send to the
+        Splunk HEC.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :param msg_timestamp: int timestamp (milliseconds) when message was sent
+          to SQS, for use as event timestamp in Splunk.
+        :type msg_timestamp: int
+        :return: list of Splunk payload dicts
+        :rtype: list
+        """
+        payloads = []
+        events = self.get_splunk_events(msg)
+        indices = self._splunk_indices_for_message(msg)
+        for event in events:
+            for index in indices:
+                payloads.append({
+                    'time': msg_timestamp,
+                    'host': 'cloud-custodian',
+                    'source': '%s-cloud-custodian' % event.get(
+                        'account', 'unknown'
+                    ),
+                    'sourcetype': '_json',
+                    'index': index,
+                    'event': event
+                })
+        return payloads
+
+    def get_splunk_events(self, msg):
+        """
+        Given a raw c7n message dict, generate a list of per-resource event
+        dictionaries to send to Splunk as events.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :return: list of Splunk HEC Event dicts
+        :rtype: list
+        """
+        # if an event is present, add it to the log
+        if msg.get('event', None) is not None:
+            # add user to the log message, at the top level...
+            user = get_aws_username_from_event(self.logger, msg['event'])
+            if user is not None:
+                msg['event_triggering_user'] = user
+        # get a copy of the message with no resources
+        base_log = deepcopy(msg)
```

Comment:
minor: deepcopy is actually fairly costly on a large message, taking a shallow copy, removing the resources, and then doing a deepcopy on the remainder would shrink the messaging being copied significantly with corresponding improvements. ie 

```python
base_log = dict(msg)
base_log.pop('resources')
base_log = deepcopy(base_log)
```